stablehlo/transforms/StablehloLegalizeQuantizedOpToQDQ.cpp

```diff
@@ -0,0 +1,135 @@
+    if (!isAnyQuantizedTypes(op->getOperandTypes()) &&
+        !isAnyQuantizedTypes(op->getResultTypes())) {
+      return failure();
+    }
+
+    SmallVector<Value> dequantizedOperands;
+    for (auto operand : op->getOperands()) {
+      if (isa<quant::QuantizedType>(getElementTypeOrSelf(operand.getType()))) {
+        dequantizedOperands.push_back(
+            rewriter.create<UniformDequantizeOp>(op->getLoc(), operand));
+      } else {
+        dequantizedOperands.push_back(operand);
+      }
+    }
+
+    auto origOp = op.getOperation();
+    auto origAttrs = origOp->getAttrs();
+    auto newOp = rewriter
+                     .create<StablehloOpType>(op.getLoc(), dequantizedOperands,
+                                              origAttrs)
+                     .getOperation();
+
+    SmallVector<Value> quantizedResults;
+    for (auto [oldResult, newResult] :
+         llvm::zip(origOp->getResults(), newOp->getResults())) {
+      if (isa<quant::QuantizedType>(
+              getElementTypeOrSelf(oldResult.getType()))) {
+        quantizedResults.push_back(
+            rewriter.create<stablehlo::UniformQuantizeOp>(
+                op->getLoc(), oldResult.getType(), newResult));
+      } else {
+        quantizedResults.push_back(newResult);
+      }
+    }
+    rewriter.replaceOp(op, quantizedResults);
+    return success();
+  }
+};
+
+class StablehloLegalizeQuantizedOpToQDQPass
+    : public impl::StablehloLegalizeQuantizedOpToQDQPassBase<
+          StablehloLegalizeQuantizedOpToQDQPass> {
+ public:
+  LogicalResult initialize(MLIRContext* context) override {
+    RewritePatternSet patterns_(context);
+    populateStablehloLegalizeQuantizedOpToQDQPatterns(&patterns_, context);
+    patterns = std::move(patterns_);
+    return success();
+  }
+
+  void runOnOperation() override {
+    auto func = getOperation();
+    if (failed(applyPatternsAndFoldGreedily(func, patterns, config))) {
+      func.emitError("Failed to converge StablehloCanonicalizeDynamism in ")
+          << config.maxIterations << " iterations";
+    }
+  }
+
+ private:
+  FrozenRewritePatternSet patterns;
+  GreedyRewriteConfig config;
+};
+
+template <typename... StablehloOpTypes>
+void populateStablehloLegalizeQuantizedOpToQDQPatterns(
+    RewritePatternSet* patterns, MLIRContext* context) {
+  patterns->add<QuantizedStablehloOpConversion<StablehloOpTypes>...>(context);
+}
+
+}  // namespace
+
+void populateStablehloLegalizeQuantizedOpToQDQPatterns(
+    RewritePatternSet* patterns, MLIRContext* context) {
+  // The following list covers most of the operations which, according to the
+  // stablehlo spoecification document, interprets the quantized
+  // operation using dequant-op-quant strategy. The ones excluded are
+  // AddOP, ConvolutionOp, DotGeneralOp, and DynamicConvOp, which are current
+  // using `stablehlo-legalize-quant-to-int` pass for decomposituion to
+  // primitive math operations.
```

Comment:
Optional: I think this pass should cover all ops. It is ok to have overlaps between this pass and stablehlo-legalize-quant-to-int. The 2 passes serve different purposes. This pass exactly matches the spec while the other tries to optimize for performance.

Reply:
Thanks for the review and approval. The main reason I kept them separate is when the passes are applied. in inder `--stablehlo-legalize-quantized-op-to-qdq --stablehlo-legalize-quant-to-int` (which is one of the use-case for converting all the quantized ops to primitive operations). If we add `AddOp` in both places then the performance optimization in later pass will never trigger. However, if we found that that the performance optimization is not desired or not in use then it should consider treating all the ops uniformly (This is something we can take up in follow up PR).